### ISSUE TITLE
Speeding up `eth_getTransactionReceipt`

### ISSUE THREAD

**Issue by fjl** [ethereum/go-ethereum#26974]
Here's an idea for speeding up RPC method `eth_getTransactionReceipt`, i.e. the operation that
loads a single receipt by txhash. Right now, this operation is very slow because it has to:

1. load inclusion block from the txhash -> block index
2. load/parse the full block from the database
3. find the tx in the block
4. verify tx signature and derive the sender
5. load/parse all receipts in the block
6. derive non-consensus fields for all receipts
7. return receipt n

Some of these steps should be avoidable. I propose that, in the tx inclusion index (txhash -> block)
we should also store the tx sender, as well as other information that required to derive receipt fields.
If this information was available, the steps for resolving a single receipts would be:

1. load the inclusion block, sender, etc. from the txhash -> block index
2. load/parse the full block from the database
3. find the tx in the block
4. load all receipts in the block
5. find receipt n
6. derive fields of the receipt
7. return it

This doesn't look very different, but there is another advantage: we wouldn't need to actually parse the receipts just to load the nth one. Decoding receipts is slow, and a bit of time can be saved by not doing it.

**Comment by fjl** [comment]
More about the additional content of the txhash -> number index. In order to derive the non-persistent fields of a single receipt, we require these new fields in the index:

- `txIndex`: the index of the transaction in the block.
- `blockLogCount`: the count of logs in the block up to this receipt. This is needed to assign the index of logs corrrectly.
- `gasUsed`: the gas used by a tx is not stored in the receipt. Instead, we store the cumulative amount of used block gas at the end of the transaction. In the index, we could simply store `gasUsed` directly. This might also be helpful for other use cases of the index.
- `txType`: the type byte of the transaction

**Comment by ameya-deshmukh** [comment]
@fjl is this still open for collaboration?

**Comment by fjl** [comment]
Yeah sure, nobody has worked on this. I just posted it as an idea.

**Comment by ameya-deshmukh** [comment]
Cool, thanks! Will start working on this

**Comment by omerfirmak** [comment]
https://github.com/ethereum/go-ethereum/pull/32063#issuecomment-2990298842

**Comment by MariusVanDerWijden** [comment]
@omerfirmak did a bunch of work to speed this up. Are more investigations necessary here or are we okay with the current state?
